File: javascript/src/diskon.ts
export enum UnitDiskon {
  PersenDariX,
  MinimumPembelianX,
  MinXMaxY,
}

/**
 *
 * @param {number} n
 * @param {UnitDiskon} unit
 * @param {number} amount
 * @param {number} x
 * @param {number} y
 * @returns {number}
 */
export function diskon(n: number, unit: UnitDiskon, amount: number, x = 0, y = 0, z = 0): number {
  if (typeof n !== 'number' && !Number.isNaN(n)) {
    throw new TypeError('isi dari argumen pertama harus berupa angka, bukan string atau lainnya');
  }

  if (typeof amount !== 'number' && !Number.isNaN(amount)) {
    throw new TypeError('isi dari argumen ketiga harus berupa angka, bukan string atau lainnya');
  }

  switch (unit) {
    case UnitDiskon.PersenDariX: {
      return n - n * (amount / 100);
    }
    case UnitDiskon.MinimumPembelianX: {
      if (y >= x) {
        return n - amount;
      }
      return n;
    }
    case UnitDiskon.MinXMaxY: {
      if (z >= x && z <= y) {
        return n - amount;
      }
      return n;
    }
    default: {
      throw new TypeError('argumen unit yang diberikan tidak valid');
    }
  }
}

File: go/diskon.go
package manusier

type DiskonType int

const (
	PersenDariX DiskonType = iota
	MinimumPembelianX
	MinXMaxY
)

// Gives discount for the given int64 argument in n with a certain unit (from supplied const enum).
// There are three ways to use this function:
//
//       import (
//         log
//         manusier "github.com/teknologi-umum/manusier/go"
//       )
//
//       func main() {
//         p := manusier.Diskon(50000, manusier.PersenDariX, 10)
//         log.Println(p) // 45000
//
//         m := manusier.Diskon(50000, manusier.MinimumPembelianX, 5000, 1, 5)
//         log.Println(p) // 45000
//
//         n := manusier.Diskon(50000, manusier.MinXMaxY, 5000, 5, 20, 10)
//         log.Println(n) // 45000
//       }
func Diskon(n int64, unit DiskonType, amount int64, xyz ...int) int64 {
	switch unit {
	case PersenDariX:
		return int64(float64(n) - (float64(n) * (float64(amount) / 100)))
	case MinimumPembelianX:
		y := xyz[1]
		x := xyz[0]
		if y >= x {
			return n - amount
		}
		return n
	case MinXMaxY:
		z := xyz[2]
		y := xyz[1]
		x := xyz[0]
		if z >= x && z <= y {
			return n - amount
		}
		return n
	default:
		panic("argumen unit yang diberikan tidak valid")
	}
}
